From a79d36a6dac6204a10ba67b7921b7918c216265f Mon Sep 17 00:00:00 2001
From: sky zhou <sky.zhou@amlogic.com>
Date: Fri, 11 Oct 2019 15:04:47 +0800
Subject: [PATCH] SurfaceFlinger: force video overlay as opaque layer [1/1]

PD#SWPL-15020

Problem:
When app set alpha with video surfaceview, we would
got blank screen.

Solution:
Force surfaceview as opaque, so sf will clean a transparent
region for video.

Verify:
verify on macroni.

Change-Id: I49790dd936a859971958253790230f403cdd8cbf
Signed-off-by: sky zhou <sky.zhou@amlogic.com>
---
 services/surfaceflinger/BufferLayer.cpp |  7 +++++++
 services/surfaceflinger/Layer.cpp       | 19 +++++++++++++++++++
 2 files changed, 26 insertions(+)

diff --git a/services/surfaceflinger/BufferLayer.cpp b/services/surfaceflinger/BufferLayer.cpp
index 9cf849a..1ee888e 100644
--- a/services/surfaceflinger/BufferLayer.cpp
+++ b/services/surfaceflinger/BufferLayer.cpp
@@ -934,6 +934,13 @@ bool BufferLayer::isOpaque(const Layer::State& s) const {
         return false;
     }
 
+#ifdef REDUCE_VIDEO_WORKLOAD
+    if (getBE().compositionInfo.mBuffer &&
+        am_gralloc_is_omx_metadata_producer(getBE().compositionInfo.mBuffer->getUsage())) {
+        return true;
+    }
+#endif
+
     // if the layer has the opaque flag, then we're always opaque,
     // otherwise we use the current buffer's format.
     return ((s.flags & layer_state_t::eLayerOpaque) != 0) || mCurrentOpacity;
diff --git a/services/surfaceflinger/Layer.cpp b/services/surfaceflinger/Layer.cpp
index 49c42af..e9ebbdb 100644
--- a/services/surfaceflinger/Layer.cpp
+++ b/services/surfaceflinger/Layer.cpp
@@ -59,6 +59,11 @@
 #include <mutex>
 #include "LayerProtoHelper.h"
 
+#ifdef REDUCE_VIDEO_WORKLOAD
+#include "OmxUtil.h"
+#include <am_gralloc_ext.h>
+#endif
+
 #define DEBUG_RESIZE 0
 
 namespace android {
@@ -1255,6 +1260,13 @@ bool Layer::setAlpha(float alpha) {
     if (mCurrentState.color.a == alpha) return false;
     mCurrentState.sequence++;
     mCurrentState.color.a = alpha;
+#ifdef REDUCE_VIDEO_WORKLOAD
+    if (getBE().compositionInfo.mBuffer &&
+        am_gralloc_is_omx_metadata_producer(getBE().compositionInfo.mBuffer->getUsage())) {
+        mCurrentState.color.a = 1.0_hf;
+    }
+#endif
+
     mCurrentState.modified = true;
     setTransactionFlags(eTransactionNeeded);
     return true;
@@ -1909,6 +1921,13 @@ Transform Layer::getTransform() const {
 half Layer::getAlpha() const {
     const auto& p = mDrawingParent.promote();
 
+#ifdef REDUCE_VIDEO_WORKLOAD
+        if (getBE().compositionInfo.mBuffer &&
+            am_gralloc_is_omx_metadata_producer(getBE().compositionInfo.mBuffer->getUsage())) {
+            return 1.0_hf;
+        }
+#endif
+
     half parentAlpha = (p != nullptr) ? p->getAlpha() : 1.0_hf;
     return parentAlpha * getDrawingState().color.a;
 }
-- 
2.7.4

